docs/config.md
# Configuring Whyis

Whyis is configured within a generated app using the file `config.py`. This includes a number of options that can be added or modified.
The file contains a baseline config object with customizations for production, development, and testing.

## Enabling Anonymous Access

By defualt, Whyis requires users to be authenticated to access the knowledge graph. 
Read-only anonymous access can be enabled by adding the following entry:

```
   DEFAULT_ANONYMOUS_READ=True,
```

## Using BlazeGraph Bulk Loading

Whyis can use the Blazegraph custom bulk loader API if configured. 
To use it with the default configuration, add the following entries to the config object:

```
    knowledge_useBlazeGraphBulkLoad = True,
    knowledge_bulkLoadEndpoint = 'http://localhost:8080/blazegraph/dataloader',
    knowledge_BlazeGraphProperties = '/apps/whyis/knowledge.properties',
    load_dir = '/data/loaded',
    knowledge_bulkLoadNamespace = 'knowledge',
    
```



docs/configuration.md
# Configure Satoru

Satoru is built on the Flask web framework, and most of the Flask authentication options are available to configure in Satoru.
The main configuration components are `config.py` and the turtle vocabulary file, at `vocab.ttl`.
`config.py` is a python-based configuration file, where the main configurations are set in a dict object.

Any knowledge graph needs to have a default URI namespace.
This namespace is the prefix for all users, nanopublications, and is the default prefix for entities in the knowledge graph.
The default prefix is for development. It should be set to the base URL of the Satoru application:

```
LOD_PREFIX='http://localhost:5000'
```

The site name should be configured as well:

```
   site_name = "Satoru Knowledge Graph",
```

You should also configure a WSRF secret key, in the `SECRET_KEY` field.
It should be set from a value generated using a command like:

```
import os; os.urandom(24)
```

Finally, set a password salt secret that will prevent hackers from being able to guess passwords if they get access to them.
This can be generated using `os.urandom(24)` as well.

```
SECURITY_PASSWORD_SALT = 'changeme__',
```

## Views and Interaction

Nodes in Satoru knowledge graphs can be customized to look like just about anything based on their `rdf:type`.
Nodes can also have multiple views, which can be selected using the `view=` URL parameter. 
The default type in the graph is `rdfs:Resource`, and the default view is `view`.
If you want to create a default view for a class you are defining, add the following to your vocab.ttl file:

```
<http://example.com/my_ontology/MyClass> rdfs:label "My Class";
    graphene:hasView "my_class_view.html".
```

`graphene:hasView` is the top-level property for creating views in Satoru.
If you want to create a new kind of view, create a subproperty of `graphene:hasView` and give it an identifier:

```
<http://example.com/my_ontology/hasScatterPlot> dc:identifier "scatter_plot";
  rdfs:subPropertyOf graphene:hasView.
```

You can now configure `scatter_plot` views for your class too:

```
<http://example.com/my_ontology/MyClass> graphene:hasScatterPlot "my_scatterplot_view.html".
```

Create a file called `/apps/myknowledgegraph/templates/my_scatterplot_view.html` and populate it with your code.
The file `/apps/satoru/templates/resource_view.html` is a good first example.

You will need to restart apache after reconfiguring views in Satoru.

The pre-configured template and views are:

| Class | view | describe | label | nanopublications | related |
| ----- | ---- | -------- | ----- | ---------------- | ------- |
|rdfs:Resource | resource_view.html | describe.json | label_view.html | nanopublications.json | related.json |
|owl:Class | class_view.html |
|owl:Ontology | ontology_view.html |
|bibo:AcademicArticle | article_view.html |

The view selector will choose the template for the view that is most specific to an instance.
Subclasses with custom views will override their transitive superclasses.
Instances do not need to have a type directly asserted, but can use the knowledge graph and vocab files to find relevant superclasses to the asserted types.

## Security
Satoru uses flask-security for authentication and authorization. 
The default configuration is to require an authenticated user for all access.
We are happy to accept pull requests that make this configurable. 
A guide to [configuring flask-security](https://pythonhosted.org/Flask-Security/configuration.html) is available on the flask-security site.

## Email
Satoru uses flask-mail for sending email. Please refer to the [flask-mail configuration guide](https://pythonhosted.org/Flask-Mail/).



config-template/{{cookiecutter.location}}/config.py
# -*- config:utf-8 -*-

import os
import logging
from datetime import timedelta

project_name = "{{cookiecutter.project_slug}}"
import importer

from whyis import autonomic
import agents.nlp as nlp
import agents.hermit as hermit
import rdflib
from datetime import datetime
from whyis.namespace import skos

# Set to be custom for your project
LOD_PREFIX = '{{cookiecutter.linked_data_prefix}}'
#os.getenv('lod_prefix') if os.getenv('lod_prefix') else 'http://hbgd.tw.rpi.edu'


from {{cookiecutter.project_slug}}.agent import *

# base config class; extend it to your needs.
Config = dict(
    # use DEBUG mode?
    DEBUG = False,

    site_name = "{{cookiecutter.project_name}}",

    site_header_image = 'static/images/random_network.png',

    site_description = '',
    
    root_path = '/apps/whyis',

    # use TESTING mode?
    TESTING = False,

    # use server x-sendfile?
    USE_X_SENDFILE = False,

    WTF_CSRF_ENABLED = True,
    SECRET_KEY = "{{cookiecutter.SECRET_KEY}}",

    nanopub_archive = {
        'depot.storage_path' : "/data/nanopublications",
    },

    file_archive = {
        'depot.storage_path' : '/data/files',
        'cache_max_age' : 3600*24*7,
    },
    vocab_file = "{{cookiecutter.location}}/vocab.ttl",
    WHYIS_TEMPLATE_DIR = [
        "{{cookiecutter.location}}/templates",
    ],
    WHYIS_CDN_DIR = "{{cookiecutter.location}}/static",

    # LOGGING
    LOGGER_NAME = "%s_log" % project_name,
    LOG_FILENAME = "/var/log/%s/output-%s.log" % (project_name,str(datetime.now()).replace(' ','_')),
    LOG_LEVEL = logging.INFO,
    LOG_FORMAT = "%(asctime)s %(levelname)s\t: %(message)s", # used by logging.Formatter

    PERMANENT_SESSION_LIFETIME = timedelta(days=7),

    # EMAIL CONFIGURATION
    ## MAIL_SERVER = "",
    ## MAIL_PORT = 587,
    ## MAIL_USE_TLS = True,
    ## MAIL_USE_SSL = False,
    ## MAIL_DEBUG = False,
    ## MAIL_USERNAME = '',
    ## MAIL_PASSWORD = '',
    ## DEFAULT_MAIL_SENDER = "{{cookiecutter.author}} <{{cookiecutter.email}}>",

    # see example/ for reference
    # ex: BLUEPRINTS = ['blog']  # where app is a Blueprint instance
    # ex: BLUEPRINTS = [('blog', {'url_prefix': '/myblog'})]  # where app is a Blueprint instance
    BLUEPRINTS = [],

    lod_prefix = LOD_PREFIX,
    SECURITY_EMAIL_SENDER = "{{cookiecutter.author}} <{{cookiecutter.email}}>",
    SECURITY_FLASH_MESSAGES = True,
    SECURITY_CONFIRMABLE = False,
    SECURITY_CHANGEABLE = True,
    SECURITY_TRACKABLE = True,
    SECURITY_RECOVERABLE = True,
    SECURITY_REGISTERABLE = True,
    SECURITY_PASSWORD_HASH = 'sha512_crypt',
    SECURITY_PASSWORD_SALT = '{{cookiecutter.SECURITY_PASSWORD_SALT}}',
    SECURITY_SEND_REGISTER_EMAIL = False,
    SECURITY_POST_LOGIN_VIEW = "/",
    SECURITY_SEND_PASSWORD_CHANGE_EMAIL = False,
    SECURITY_DEFAULT_REMEMBER_ME = True,
    ADMIN_EMAIL_RECIPIENTS = [],
    db_defaultGraph = LOD_PREFIX + '/',

    java_classpath = "{{cookiecutter.location}}/jars",

    admin_queryEndpoint = 'http://localhost:8080/blazegraph/namespace/admin/sparql',
    admin_updateEndpoint = 'http://localhost:8080/blazegraph/namespace/admin/sparql',
    
    knowledge_queryEndpoint = 'http://localhost:8080/blazegraph/namespace/knowledge/sparql',
    knowledge_updateEndpoint = 'http://localhost:8080/blazegraph/namespace/knowledge/sparql',

    LOGIN_USER_TEMPLATE = "auth/login.html",
    CELERY_BROKER_URL = 'redis://localhost:6379/0',
    CELERY_RESULT_BACKEND = 'redis://localhost:6379/0',
    default_language = 'en',
    namespaces = [
        importer.LinkedData(
            prefix = LOD_PREFIX+'/doi/',
            url = 'http://dx.doi.org/%s',
            headers={'Accept':'text/turtle'},
            format='turtle',
            postprocess_update= '''insert {
                graph ?g {
                    ?pub a <http://purl.org/ontology/bibo/AcademicArticle>.
                }
            } where {
                graph ?g {
                    ?pub <http://purl.org/ontology/bibo/doi> ?doi.
                }
            }
            '''
        ),
        importer.LinkedData(
            prefix = LOD_PREFIX+'/dbpedia/',
            url = 'http://dbpedia.org/resource/%s',
            headers={'Accept':'text/turtle'},
            format='turtle',
            postprocess_update= '''insert {
                graph ?g {
                    ?article <http://purl.org/dc/terms/abstract> ?abstract.
                }
            } where {
                graph ?g {
                    ?article <http://dbpedia.org/ontology/abstract> ?abstract.
                }
            }
            '''
        )
    ],
    inferencers = {
        "SETLr": autonomic.SETLr(),
        "Consistency Check" : hermit.ConsistencyCheck()
#        "HTML2Text" : nlp.HTML2Text(),
#        "EntityExtractor" : nlp.EntityExtractor(),
#        "EntityResolver" : nlp.EntityResolver(),
#        "TF-IDF Calculator" : nlp.TFIDFCalculator(),
#        "SKOS Crawler" : autonomic.Crawler(predicates=[skos.broader, skos.narrower, skos.related])
    },
    inference_tasks = [
#        dict ( name="SKOS Crawler",
#               service=autonomic.Crawler(predicates=[skos.broader, skos.narrower, skos.related]),
#               schedule=dict(hour="1")
#              )
    ],

    base_rate_probability = 0.5
)


# config class for development environment
Dev = dict(Config)
Dev.update(dict(
    DEBUG = True,  # we want debug level output
    MAIL_DEBUG = True,
    EXPLAIN_TEMPLATE_LOADING = True,
    DEBUG_TB_INTERCEPT_REDIRECTS = False
))

# config class used during tests
Test = dict(Config)
Test.update(dict(
    TESTING = True,
    WTF_CSRF_ENABLED = False
))

